Improved unit test coverage of mig lib daemon

jonasbardino opened the pull request:
Improved unit test coverage of `mig/lib/daemon.py` for complete coverage. Added a lot of corner case tests to test signal passing and handlers. Some of the tests include fuzzy timing assertions, which _may_ need further tweaking to avoid false positives.

Adds one new function to the tested library to remove already registered signal handlers, but it is only ever used in the unit tests so far.
Slightly tweaks `interruptible_sleep` corner cases to fail an assertion if  called with non-positive `nap_secs` and to return right away if called with `max_sleep` less than `nap_secs`. Both are unused corner cases only used in the test for now but they may or may not be of use to daemons in the future.

jonasbardino committed:
Refactor to use a shared before_each init to isolate test configuration etc.
Add coverage of several previously missing branches and corner cases.
- Direct event helper testing
- Signal handler verification
- Concurrent event handling
- Nap timing accuracy checks
- Multiple reset cycles
- Error cases like invalid arguments
- Various sleep interruption scenarios

jonasbardino committed:
Implement a new helper to unregister all registered signals and use it in unit
tests to increase isolation and be able to test handling when no handlers where
registered.

Added unit test coverage:
 - Immediate sleep exit with max_secs < nap_secs
 - Positional argument handling
 - Negative max_secs handling
 - Zero max_secs handling
 - Unregistered signal safety
 - Consecutive signal handling

Perhaps a stand-alone test for the new unregister_signal_handlers should also
be added.

jonasbardino committed:
Extended daemon module coverage further:
 - Verify explicit unregistration of signal handlers
 - Test sleep terminating after a condition becomes true
 - Edge case test for equal max/sleep durations
 - Test error handling when break conditions raise exceptions
 - Direct tests of state reset and setting helpers
 - Integration test with real signal delivery

albu-diku commented on tests/test_mig_lib_daemon.py:
```diff
@@ -27,27 +27,44 @@
+
+        # Create fake configuration, sig and frame for test isolation
+        self.dummy_conf = FakeConfiguration()
+        self.dummy_conf.logger = FakeLogger()
```
Out test case objects give you a configured FakeLogger via `self.logger` - those are also wired into the logger checking mechanism and thus will report spurious errors.

jonasbardino replied:
Ah, I see - I'll switch.

albu-diku commented on tests/test_mig_lib_daemon.py:
```diff
@@ -27,27 +27,44 @@
+        reset_stop()
+
+        # Unregister any existing signal handlers
+        used_signals = [signal.SIGCONT, signal.SIGINT, signal.SIGALRM,
```
This array can be defined once at the top of the file rather than be recreated on each `before_each` call. It might be worth defining a function called `_unregister_all_signals()` and just calling that.

Question on my part, what is it in signal unregistration that depends on a configuration argument?

jonasbardino replied:
Yes, it could. I'll consider if it's worth it with another `unregister` wrapper.

There's no hard need for `configuration` in `unregister` or the symmetric `register` so that is mostly a gut feeling that it may come in handy at some point and it is a pain to add later.

jonasbardino commented on tests/test_mig_lib_daemon.py:
```diff
@@ -27,27 +27,44 @@
+        """Set up test configuration and reset state before each test"""
+
+        # Create fake configuration, sig and frame for test isolation
+        self.dummy_conf = FakeConfiguration()
```
Alright, it looks like we started to diverge on attribute naming. I'll try to re-sync.

jonasbardino committed:
Switch dummy conf and logger to the default ones from `_provide_configuration`.
Move used_signal accounting to class-level. Perhaps we want to maintain it on
actual use and add helpers to track active ones eventually.

jonasbardino committed:
Address the review comment on PR367 about the edge cases not really being
verified by inserting missing `assert`s.

jonasbardino commented on tests/test_mig_lib_daemon.py:
```diff
@@ -103,12 +115,368 @@ def test_register_stop_handler_signal(self):
+        self.assertTrue(time.time() - start < 0.05)
+
+        # Test zero and negative max_secs returns immediately
+        interruptible_sleep(self.configuration, 0.0, [])
```
It indeed looks like the corresponding asserts went missing. Added them now.

The idea of adjusting the return value of `interruptible_sleep` to hint about actual time slept is not bad :+1:  Not needed as-is but could be a future target - in a separate PR to keep code changes and unit tests separate.

jonasbardino committed:
Adjust coflicting handler test to behave as expected with only the last handler
assigned being active.

jonasbardino committed:
Increase test coverage to 100% with more corner case tests of sleep intervals,
unregister, etc.

jonasbardino committed:
Remove useless and unused result from `interruptible_sleep`, also called out by
pylint.

albu-diku reviewed (approved):
After the last set of small tweaks I think we benefit from this being in-tree as found.
